File: docs/lab1alt/code/Main.java
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;
import java.io.IOException;

public class Main {
    public static void main(String[] args) throws IOException {
        CharStream  input  = CharStreams.fromStream(System.in);
        Lexer       lexer  = new ExprLexer(input);
        TokenStream tokens = new CommonTokenStream(lexer);
        ExprParser  parser = new ExprParser(tokens);
        ParseTree   tree   = parser.expr(); // 取得一颗以 expr 为根的 AST
        System.out.println(tree.toStringTree(parser));
    }
}

File: docs/lab1alt/code/main.py
import sys
from antlr4 import *
from ExprLexer import ExprLexer
from ExprParser import ExprParser


def main():
    input  = InputStream(sys.stdin.read())
    lexer  = ExprLexer(input)
    tokens = CommonTokenStream(lexer)
    parser = ExprParser(tokens)
    tree   = parser.expr() # 取得一颗以 expr 为根的 AST
    print(tree.toStringTree(recog=parser))


if __name__ == '__main__':
    main()

